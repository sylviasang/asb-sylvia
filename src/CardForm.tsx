import React from 'react';
import styled from 'styled-components';
import { Button, IconButton, makeStyles } from '@material-ui/core';
import MenuIcon from '@material-ui/icons/Menu';
import TextField from '@material-ui/core/TextField';

const Container = styled.div`
    display: flex;
    flex-direction: column
`;

const NavigationContainer = styled.div`
    display: flex;
    flex-direction: row;
    border-bottom: 1px solid #000000;
`;

const Header = styled.div`
    width: 100%;
    display: flex;
    justify-content: center;
`;

const FormContainer = styled.div`
    display: flex;
    justify-content: center;
    margin-left: 180px;
`;

const Form = styled.div`
    height: calc(100vh - 150px);
    display: flex;
    flex-direction: column;
    justify-content: space-between;
`;

const CardForm:React.FC = () => {
  return (
    <Container>
        <NavigationContainer>
            <IconButton name='HamburgerButton' data-testid='HamburgerButton'><MenuIcon fontSize='large'/></IconButton>
            <Header><h1 data-testid='Register'>Register Card Form</h1></Header>
        </NavigationContainer>
        <FormContainer>
            <Form>
                <h2 style={{marginTop: '20px'}}>Welcome {'{'} User.FirstName {'}'}</h2>
                <form>
                    <TextField
                        name='Credit Card Number Input'
                        data-testid='CreditCardNumber'
                        style={{display: 'block', marginBottom: '20px', width: 400}} 
                        required 
                        label="Credit Card Number" 
                        variant="outlined"
                    />
                    <TextField 
                        name='CVC'
                        data-testid='CVC'
                        style={{width: '100px', marginRight: '20px'}}
                        required 
                        label="CVC" 
                        variant="outlined" 
                    />
                    <TextField
                        name='Expire'
                        data-testid='Expire'
                        style={{width: '120px', marginRight: '20px'}}
                        required 
                        label="Expire" 
                        variant="outlined" 
                        /> 
                    <Button
                        name='Submit'
                        data-testid='Submit'
                        variant="contained" 
                        type="submit" 
                        size="large" 
                        style={{display: 'block', marginTop: '20px', width: 200}}>
                        Submit
                    </Button>
                </form>
            </Form>
        </FormContainer>
    </Container>

  );
}

export default CardForm;
