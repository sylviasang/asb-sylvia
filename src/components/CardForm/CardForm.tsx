import React, {useState} from 'react';
import styled from 'styled-components';
import { Button, IconButton, makeStyles } from '@material-ui/core';
import MenuIcon from '@material-ui/icons/Menu';
import TextField from '@material-ui/core/TextField';

const Container = styled.div`
    display: flex;
    flex-direction: column;
`;

const NavigationContainer = styled.div`
    display: flex;
    flex-direction: row;
    border-bottom: 1px solid #000000;
`;

const Header = styled.div`
    width: 100%;
    display: flex;
    justify-content: center;
`;

const FormContainer = styled.div`
    display: flex;
    justify-content: center;
    margin-left: 180px;
`;

const Form = styled.div`
    height: calc(100vh - 150px);
    display: flex;
    flex-direction: column;
    justify-content: space-between;
`;

interface CardFormProps {
    handleClick: any
};

const CardForm:React.FC<CardFormProps> = (props) => {

  const [creditCardNo, setCreditCardNo] = useState<number>();
  const [cvc, setCvc] = useState<number>();
  const [expire, setExpire] = useState<Date>();
  const [creditCardError, setCreditCardError] = useState<boolean>(false);
  const [creditCardErrorInfo, setCreditCardErrorInfo] = useState<string>('');
  const [cvcError, setCvcError] = useState<boolean>(false);
  const [cvcErrorInfo, setCvcErrorInfo] = useState<string>('');
  const [expireError, setExpireError] = useState<boolean>(false);
  const [expireErrorInfo, setExpireErrorInfo] = useState<string>('');

  const handleCreditCardChange = (event: any) => {
    const creditCard = event.target.value;
    setCreditCardNo(creditCard);
    if (creditCard && !Number(creditCard)) {
        setCreditCardError(true);
        setCreditCardErrorInfo('This field should be a number');
    } else {
        setCreditCardError(false)
        setCreditCardErrorInfo('');
    }
  }

  const handleCvcChange = (event: any) => {
    const cvc = event.target.value;
    setCvc(cvc);
    if (cvc && !Number(cvc)) {
        setCvcError(true);
        setCvcErrorInfo('This field should be a number');
    } else {
        setCvcError(false);
        setCvcErrorInfo('');
    }
  }
  const handleExpireChange = (event: any) => {
    const expire = event.target.value;
    const regx = /^(\d{4})-(\d{2})-(\d{2})$/
    setExpire(expire);
    if (expire && !regx.test(expire)) {
        setExpireError(true);
        setExpireErrorInfo('This field should be a valid date e.g. 2021-08-29');
    } else {
        setExpireError(false)
        setExpireErrorInfo('');
    }
  }
  const handleSubmit = (event:any) => {
    event.preventDefault();

    if (creditCardError || cvcError || expireError ){
        alert('please correct the input fields')
        return
    }
    console.log(
        'credit card number: ', {creditCardNo},
        'cvc: ', {cvc},
        'expire', {expire},
    )
  }

  return (
    <Container>
        <NavigationContainer>
            <IconButton name='HamburgerButton' data-testid='HamburgerButton' onClick={() => props.handleClick()}><MenuIcon fontSize='large'/></IconButton>
            <Header><h1 data-testid='Register'>Register Card Form</h1></Header>
        </NavigationContainer>
        <FormContainer>
            <Form>
                <h2 style={{marginTop: '20px'}}>Welcome {'{'} User.FirstName {'}'}</h2>
                <form onSubmit={handleSubmit}>
                    <TextField
                        name='Credit Card Number Input'
                        data-testid='CreditCardNumber'
                        style={{display: 'block', marginBottom: '20px', width: 400}} 
                        required 
                        label="Credit Card Number" 
                        variant="outlined" 
                        value={creditCardNo} 
                        onChange={handleCreditCardChange}
                        error={creditCardError}
                        helperText={creditCardErrorInfo}
                    />
                    <TextField 
                        name='CVC'
                        data-testid='CVC'
                        style={{width: '100px', marginRight: '20px'}}
                        required 
                        label="CVC" 
                        variant="outlined" 
                        value={cvc} 
                        onChange={handleCvcChange}
                        error={cvcError}
                        helperText={cvcErrorInfo}                
                        
                    />
                    <TextField
                        name='Expire'
                        data-testid='Expire'
                        style={{width: '120px', marginRight: '20px'}}
                        required 
                        label="Expire" 
                        variant="outlined" 
                        value={expire} 
                        onChange={handleExpireChange}
                        error={expireError}
                        helperText={expireErrorInfo}   
                        
                        /> 
                    <Button
                        name='Submit'
                        data-testid='Submit'
                        variant="contained" 
                        type="submit" 
                        size="large" 
                        style={{display: 'block', marginTop: '20px', width: 200}}>
                        Submit
                    </Button>
                </form>
            </Form>
        </FormContainer>
    </Container>

  );
}

export default CardForm;
