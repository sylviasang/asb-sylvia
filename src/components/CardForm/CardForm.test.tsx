import { render, screen, fireEvent, getByTestId, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import React from 'react';
import CardForm from './CardForm';

afterEach(() => {
    cleanup();
})

test('render with props', () => {
    const onClick = jest.fn();
    const { container } = render(<CardForm handleClick={onClick}/>);
    expect(container.firstChild).toMatchSnapshot();
})

test('render when hamburger button click', () => {
    const onClick = jest.fn();
    const { container } = render(<CardForm handleClick={onClick}/>);
    fireEvent.click(screen.getByTestId('HamburgerButton'));
    expect(onClick).toHaveBeenCalled();
})

test('test credit card input', () => {
    const onClick = jest.fn();
    const { container } = render(<CardForm handleClick={onClick}/>);
    const creditCardInput = screen.getByTestId('CreditCardNumber');
    
    fireEvent.change(creditCardInput, {target: {value: 123456789}});

    expect(creditCardInput.value).toBe('123456789')
    expect(screen.queryByText(/This field should be a number/)).toBeNull();

    fireEvent.change(creditCardInput, {target: {value: 'CreditCardNumber'}});

    expect(creditCardInput.value).toBe('CreditCardNumber')
    expect(screen.queryByText(/This field should be a number/)).toBeInTheDocument();
})

test('test cvc input', () => {
    const onClick = jest.fn();
    const { container } = render(<CardForm handleClick={onClick}/>);
    const cvcInput = screen.getByTestId('CVC');
    
    fireEvent.change(cvcInput, {target: {value: 123456789}});

    expect(cvcInput.value).toBe('123456789')
    expect(screen.queryByText(/This field should be a number/)).toBeNull();

    fireEvent.change(cvcInput, {target: {value: 'CreditCardNumber'}});

    expect(cvcInput.value).toBe('CreditCardNumber')
    expect(screen.queryByText(/This field should be a number/)).toBeInTheDocument();

})

test('test expire input', () => {
    const onClick = jest.fn();
    const { container } = render(<CardForm handleClick={onClick}/>);
    const expireInput = screen.getByTestId('Expire');
    
    fireEvent.change(expireInput, {target: {value: '2021-08-29'}});

    expect(expireInput.value).toBe('2021-08-29')
    expect(screen.queryByText(/This field should be a valid date e.g. 2021-08-29/)).toBeNull();

    fireEvent.change(expireInput, {target: {value: 'CreditCardNumber'}});

    expect(expireInput.value).toBe('CreditCardNumber')
    expect(screen.queryByText(/This field should be a valid date e.g. 2021-08-29/)).toBeInTheDocument();
})

test('render when submit button click with uncorrct input values', () => {
    const onClick = jest.fn();
    const { container } = render(<CardForm handleClick={onClick}/>);
    const alertMock = jest.spyOn(window,'alert').mockImplementation(); 

    fireEvent.change(screen.getByTestId('CreditCardNumber'), {target: {value: 'non number input'}});
    fireEvent.change(screen.getByTestId('CVC'), {target: {value: 'non number input'}})
    fireEvent.change(screen.getByTestId('Expire'), {target: {value: 'non date input'}})
    fireEvent.click(screen.getByTestId('Submit'));

    expect(alertMock).toHaveBeenCalledTimes(1);

})
