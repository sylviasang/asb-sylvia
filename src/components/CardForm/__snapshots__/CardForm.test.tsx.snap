// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render with props 1`] = `
<div
  class="sc-bdfBQB aujAO"
>
  <div
    class="sc-gsTEea hGQBnA"
  >
    <button
      class="MuiButtonBase-root MuiIconButton-root"
      data-testid="HamburgerButton"
      name="HamburgerButton"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiIconButton-label"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M3 18h18v-2H3v2zm0-5h18v-2H3v2zm0-7v2h18V6H3z"
          />
        </svg>
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <div
      class="sc-dlfnuX jzRvPj"
    >
      <h1
        data-testid="Register"
      >
        Register Card Form
      </h1>
    </div>
  </div>
  <div
    class="sc-hKgJUU hZthbb"
  >
    <div
      class="sc-eCstlR hAHCGk"
    >
      <h2
        style="margin-top: 20px;"
      >
        Welcome 
        {
         User.FirstName 
        }
      </h2>
      <form>
        <div
          class="MuiFormControl-root MuiTextField-root"
          style="display: block; margin-bottom: 20px; width: 400px;"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
            data-shrink="false"
          >
            Credit Card Number
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              data-testid="CreditCardNumber"
              name="Credit Card Number Input"
              required=""
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-3"
              >
                <span>
                  Credit Card Number
                   *
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root"
          style="width: 100px; margin-right: 20px;"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
            data-shrink="false"
          >
            CVC
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              data-testid="CVC"
              name="CVC"
              required=""
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-3"
              >
                <span>
                  CVC
                   *
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
        <div
          class="MuiFormControl-root MuiTextField-root"
          style="width: 120px; margin-right: 20px;"
        >
          <label
            class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
            data-shrink="false"
          >
            Expire
            <span
              aria-hidden="true"
              class="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          >
            <input
              aria-invalid="false"
              class="MuiInputBase-input MuiOutlinedInput-input"
              data-testid="Expire"
              name="Expire"
              required=""
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="PrivateNotchedOutline-root-1 MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="PrivateNotchedOutline-legendLabelled-3"
              >
                <span>
                  Expire
                   *
                </span>
              </legend>
            </fieldset>
          </div>
          
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSizeLarge MuiButton-sizeLarge"
          data-testid="Submit"
          name="Submit"
          style="display: block; margin-top: 20px; width: 200px;"
          tabindex="0"
          type="submit"
        >
          <span
            class="MuiButton-label"
          >
            Submit
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </div>
</div>
`;
